src/core/src/pattern/op/optional.cpp

```diff
@@ -0,0 +1,46 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "openvino/pass/pattern/op/optional.hpp"
+
+#include "openvino/pass/pattern/matcher.hpp"
+
+std::unordered_set<ov::DiscreteTypeInfo> ov::pass::pattern::op::Optional::get_optional_types() const {
+    return optional_types;
+}
+
+bool ov::pass::pattern::op::Optional::match_value(Matcher* matcher,
+                                                  const Output<Node>& pattern_value,
+                                                  const Output<Node>& graph_value) {
+    const auto p_value = input_value(0);
```

Comment:
I believe we can re-use WrapType logic here and just add sole "else" branch
if ( // code from WrapType class, "match_value" method)
else { 
   return matcher->match_value( parent_of_the_current_pattern_node, the_same_graph_value)
}